pkg/seedadmissioncontroller/webhooks/admission/extensionresources/admission.go

```diff
@@ -0,0 +1,252 @@
+				return validation.ValidateOperatingSystemConfig(n.(*extensionsv1alpha1.OperatingSystemConfig))
+			},
+			validateUpdateResource: func(n, o client.Object) field.ErrorList {
+				return validation.ValidateOperatingSystemConfigUpdate(n.(*extensionsv1alpha1.OperatingSystemConfig), o.(*extensionsv1alpha1.OperatingSystemConfig))
+			},
+		},
+
+		gvr("workers"): &artifact{
+			newObject: func() client.Object { return new(extensionsv1alpha1.Worker) },
+			validateCreateResource: func(n, _ client.Object) field.ErrorList {
+				return validation.ValidateWorker(n.(*extensionsv1alpha1.Worker))
+			},
+			validateUpdateResource: func(n, o client.Object) field.ErrorList {
+				return validation.ValidateWorkerUpdate(n.(*extensionsv1alpha1.Worker), o.(*extensionsv1alpha1.Worker))
+			},
+		},
+	}
+
+	h := handler{
+		logger:    logger,
+		artifacts: artifacts,
+	}
+
+	return &h
+}
+
+type handler struct {
+	decoder   *admission.Decoder
+	logger    logr.Logger
+	artifacts map[metav1.GroupVersionResource]*artifact
+}
+
+var _ admission.Handler = &handler{}
+
+func (h *handler) InjectDecoder(d *admission.Decoder) error {
+	h.decoder = d
+	return nil
+}
+
+func (h *handler) Handle(ctx context.Context, request admission.Request) admission.Response {
+	_, cancel := context.WithTimeout(ctx, 10*time.Second)
+	defer cancel()
+
+	artifact, ok := h.artifacts[request.Resource]
+	if !ok {
+		return admission.Allowed("validation not found for the given resource")
+	}
+
+	switch request.Operation {
+	case admissionv1.Create:
+		return h.handleValidation(request, artifact.newObject, artifact.validateCreateResource)
+	case admissionv1.Update:
+		return h.handleValidation(request, artifact.newObject, artifact.validateUpdateResource)
+	default:
+		return admission.Allowed("operation is not CREATE or UPDATE")
+	}
+}
+
+type (
+	newObjectFunc func() client.Object
+	validateFunc  func(new, old client.Object) field.ErrorList
+)
+
+// artifact servers as a helper to setup the corresponding function.
+type artifact struct {
+	// newObject is a simple function that creates and returns a new resource.
+	newObject newObjectFunc
+
+	// validateCreateResource is a wrapper function for the different create validation functions.
+	validateCreateResource validateFunc
+
+	// validateUpdateResource is a wrapper function for the different update validation functions.
+	validateUpdateResource validateFunc
+}
+
+func (h handler) handleValidation(request admission.Request, newObject newObjectFunc, validate validateFunc) admission.Response {
+	obj := newObject()
+	if err := h.decoder.DecodeRaw(request.Object, obj); err != nil {
+		h.logger.Error(err, "could not decode object", "object", request.Object)
+		return admission.Errored(http.StatusUnprocessableEntity, fmt.Errorf("could not decode object %v: %w", request.Object, err))
+	}
+
+	h.logger.Info("Validating resource", "type", request.Resource, "operation", request.Operation, "kind", obj.GetObjectKind(), "object", kutil.ObjectName(obj))
```

Comment:
Resource and kind contain the same information; resource and name should follow in this order. Also, please move the log to just before calling the `validate` function.
```suggestion
	h.logger.Info("Validating extension resource", "resource", request.Resource, "name", kutil.ObjectName(obj), "operation", request.Operation)
```